From bc45a49360485b1d196253a48655d50645b7c831 Mon Sep 17 00:00:00 2001
From: Balamurugan S <quic_bselvara@quicinc.com>
Date: Tue, 31 Jan 2023 00:09:11 +0530
Subject: [PATCH 2/8] ath12k: ath12k: Move ath12k module initialization to
 core.c

Move ath12k module initialization to core.c file. PCIe and AHB
specific inits and functionality will be in respective driver files and both
driver is initialized from the ath12k module.

Signed-off-by: Balamurugan S <quic_bselvara@quicinc.com>
---
 drivers/net/wireless/ath/ath12k/core.c | 47 +++++++++++++++++++++++++-
 drivers/net/wireless/ath/ath12k/core.h | 15 ++++++++
 drivers/net/wireless/ath/ath12k/pci.c  | 20 ++---------
 3 files changed, 63 insertions(+), 19 deletions(-)

diff --git a/drivers/net/wireless/ath/ath12k/core.c b/drivers/net/wireless/ath/ath12k/core.c
index 15fe158..e54f124 100644
--- a/drivers/net/wireless/ath/ath12k/core.c
+++ b/drivers/net/wireless/ath/ath12k/core.c
@@ -1840,5 +1840,50 @@ err_sc_free:
 	return NULL;
 }
 
-MODULE_DESCRIPTION("Core module for Qualcomm Atheros 802.11be wireless LAN cards.");
+static int ath12k_init(void)
+{
+	int ret;
+
+	ret = ath12k_debugfs_create();
+	if (ret) {
+		pr_err("Failed to register debugfs ath12k driver: %d\n", ret);
+		return ret;
+	}
+
+	ret = ath12k_ahb_init();
+	if (ret) {
+		ret = -ENODEV;
+		pr_err("Failed to initialize ath12k AHB device: %d\n", ret);
+		goto error;
+	}
+
+	ret = ath12k_pci_init();
+	if (ret) {
+		ret = -ENODEV;
+		pr_err("Failed to initialize ath12k PCI device: %d\n", ret);
+		goto error_ahb;
+	}
+
+	return 0;
+
+error_ahb:
+	ath12k_ahb_exit();
+error:
+	ath12k_debugfs_destroy();
+
+	return ret;
+}
+
+static void ath12k_exit(void)
+{
+	ath12k_debugfs_destroy();
+	ath12k_ahb_exit();
+	ath12k_pci_exit();
+
+}
+
+module_init(ath12k_init)
+module_exit(ath12k_exit)
+
+MODULE_DESCRIPTION("Driver support for Qualcomm Technologies 802.11be WLAN PCIe/AHB devices");
 MODULE_LICENSE("Dual BSD/GPL");
diff --git a/drivers/net/wireless/ath/ath12k/core.h b/drivers/net/wireless/ath/ath12k/core.h
index 8dc96ef..8a34d29 100644
--- a/drivers/net/wireless/ath/ath12k/core.h
+++ b/drivers/net/wireless/ath/ath12k/core.h
@@ -1404,4 +1404,19 @@ static inline const char *ath12k_bus_str(enum ath12k_bus bus)
 
 	return "unknown";
 }
+
+int ath12k_pci_init(void);
+void ath12k_pci_exit(void);
+#ifdef CPTCFG_ATH12K_AHB
+int ath12k_ahb_init(void);
+void ath12k_ahb_exit(void);
+#else
+static inline int ath12k_ahb_init(void)
+{
+	return 0;
+};
+
+static inline void ath12k_ahb_exit(void) {};
+#endif
+
 #endif /* _CORE_H_ */
diff --git a/drivers/net/wireless/ath/ath12k/pci.c b/drivers/net/wireless/ath/ath12k/pci.c
index 204f9f4..f473f15 100644
--- a/drivers/net/wireless/ath/ath12k/pci.c
+++ b/drivers/net/wireless/ath/ath12k/pci.c
@@ -1450,36 +1450,20 @@ static struct pci_driver ath12k_pci_driver = {
 #endif
 };
 
-static int ath12k_pci_init(void)
+int ath12k_pci_init(void)
 {
 	int ret;
 
-	ret = ath12k_debugfs_create();
-	if (ret)
-		pr_info("failed to register debugfs ath12k driver: %d\n", ret);
-
 	ret = pci_register_driver(&ath12k_pci_driver);
 	if (ret) {
 		pr_err("failed to register ath12k pci driver: %d\n",
 		       ret);
-		goto err;
 	}
 
-	return 0;
-
-err:
-	ath12k_debugfs_destroy();
 	return ret;
 }
-module_init(ath12k_pci_init);
 
-static void ath12k_pci_exit(void)
+void ath12k_pci_exit(void)
 {
 	pci_unregister_driver(&ath12k_pci_driver);
-	ath12k_debugfs_destroy();
 }
-
-module_exit(ath12k_pci_exit);
-
-MODULE_DESCRIPTION("Driver support for Qualcomm Technologies 802.11be WLAN PCIe devices");
-MODULE_LICENSE("Dual BSD/GPL");
-- 
2.17.1

